### ISSUE TITLE
Stitch multiple images to a single one problem (linked to issue #526)

### ISSUE THREAD

**Issue by alonit** [jcupitt/libvips#530]
Further to issue #526, I checked vips_arrayjoin(...) and get a single tiled image output where the edges of the stitched images can be seen clearly so that they are not smooth. I have overlap of about 5% between each neighbor images (both in x and y axis) and this duplication is seen also in the stitched image. I use it as follows:

vips_arrayjoin(images,

```
        &ImOutTotal,
        IMAGES,
        "across", 26,
        NULL)
```

Where `images`is an array of 1508 `VipsImage`, in a grid of 26 by 58.

I saw that there are another VIPS C API functions to get smooth seam image such as vips_merg (), vips_mosaic(), vips_mosaic1(), vips_match(), vips_globalbalance(), vips_remosaic() .

So, my question is: What is the more accurate and efficient way to create smooth seam, stitched images? `merge`? `join` ? `mosaic` ? `globalbalance`? given that:
1) input images are overlapped in x and y axis's
2)neighbor images within given grid row have position offset in y axis (probably due to imag scanner HW curvature)
3) there is contrast differences between input images and along the seams that I need to fix
4) it is best to get  feathered edges (smooth seam) stitched image output.

Thanks

**Comment by alonit** [comment]
Hi,

I tried `vips_mosaic` on 2 .jpg images called 1.jpg and 2.jpg ,2356x1772 pixels each where 1.jpg is the left side image and 2.jpg is the right side image. The purpose is to get one seamless image.

I read both images:

```
Im1 = vips_image_new_from_file("1.jpg",
			"access", VIPS_ACCESS_SEQUENTIAL_UNBUFFERED,
			NULL)

Im2 = vips_image_new_from_file("2.jpg",
			"access", VIPS_ACCESS_SEQUENTIAL_UNBUFFERED,
			NULL)
```

Pixel (2337,567) in 1.jpg matches pixel (27,549) in 2.jpg (roughly), so I called the following VIPS C API:

`vips_mosaic(Im1, Im2, &ImOut, VIPS_DIRECTION_HORIZONTAL, 2337, 567, 27, 549, NULL);`

and get error: 

`0x000000006a01cac0 "VipsJpeg: out of order read at line 1772\nVipsJpeg: out of order read at line 1772\nVipsJpeg: out of order read at line 1772\n"`

Thanks again.

**Comment by jcupitt** [comment]
Hi, you can't use seq images on mosaic, try just `vips_image_new_from_file("1.jpg", NULL);`.

Do you know the exact overlap between your images, or do you want to search for tie-points? Do your images have changes of scale or rotation, or will a simple translate work? Does your image capture device have a linear response, or is there something like a gamma (for example)? If you are moving a camera to take the frames, you'll find you need to correct for scatter in the optics, and perhaps pincushion or barrel distortion. Does the black point shift around? Do you move the lights? How flat is your object? Is the pane of motion of the camera parallel to the object plane? It can be complicated to make a nice mosaic :( 

The `mosaic1` functions will join a pair of images with a search to find the exact overlap. The `mosaic2` ones will find a rotation and scale as well. The `merge` functions are if you know the exact overlap in advance. `global_balance` can remove contrast differences, but you do need to model your camera response.

**Comment by alonit** [comment]
Hi, I'm working with default `vips_image_new_from_file("1.jpg", NULL);` in the last couple of days and it works for me with all the mosaicing functions `(merge , joint , mosaic . ...)` so thank you for this suggestion.  Yet, I'm concern about the memory usage in this mode as I have to mosaic about 100x100 images at max (each 2536x1772 pixels).

I'm also learned the various mosaicing functions behavior (their input/output and functionality) and it seems that I need to use `vips_mosaic` version because the followings:
1) I don't know the exact overlap but just approximation in percents (2% overlap, 3% overlap and so on)
2) All input images have the same scale and rotation.
3) All input images have gamma 1 (for now), but the background is still not uniform so I found `global_balance` useful to fix this.
4) For each frame capture, all involved scanner HW are stalled, so no scatters or distortions presented because of moving parts.
5) The pane of motion of the camera is parallel to the object plane
6) My objects are composed from tissue. Tissue thickness is about 3-5 micron, but between neighbor images there is not much difference.

Thanks

**Comment by jcupitt** [comment]
Yes, 100x100 is probably too large. I would build in sections, perhaps 20x20, then assemble those as a 5x5. Use vips format files for the intermediates. Building in two stages will also help distribute mosaicing errors.

Is this a microscope + translation stage? You might need to correct for vignetting and lens distortion too.

**Comment by alonit** [comment]
My images (1280x1024 pixels) have 1.6% overlap in x axis. When calling `vips_mosaic`:

I get error: 
`im_lrmosaic: overlap too small for search `

Is there a minimum overlap value for this function ?

Thanks

**Comment by jcupitt** [comment]
Yes, it needs to place a window of size `hwindow * 2 + 1` into the overlap for the search. Turn `hwindow` and `harea` down a bit. 

http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/libvips-mosaicing.html#vips-mosaic

**Comment by alonit** [comment]
I use the default size of 1 for both `hwindow `and `harea ` and get:

`im_lrmosaic: overlap too small for search`

 Trying to set the minimum of 0 for them return with crash or an error.

Is there a document that explain the meaning of `hwindow `and `harea `and their responsibilities in mosaic functionality - I think it would be very helpful.

Thanks

**Comment by alonit** [comment]
As I understand from the above link:

```
int
vips_mosaic (VipsImage *ref,
             VipsImage *sec,
             VipsImage **out,
             VipsDirection direction,
             int xref,
             int yref,
             int xsec,
             int ysec,
             ...);
```
Pixel (xref,yref) in `ref `image is approximatly the pixel (xsec, ysec) in `sec `image - Is that right?
So, how I configure the overlap area and the search area for this `vips_mosaic `function?

Many thanks

**Comment by jcupitt** [comment]
It's in the docs, you need:

http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/libvips/libvips-mosaicing.html#vips-mosaic

I think you have about a 20 pixel overlap, is that right? I experimented and this seems to work for me:

```
$ vips crop wtc.jpg left.v 0 0 1280 1024
$ vips crop wtc.jpg right.v 1260 0 1280 1024
$ vips mosaic left.v right.v join.v horizontal 1270 0 10 0 --harea 6 --hwindow 3
```

**Comment by jcupitt** [comment]
From C you set options like this:

```
if (vips_mosaic(left, right, &out, VIPS_DIRECTION_HORIZONTAL, 1270, 0, 10, 0,
    "harea", 6,
    "hwindow", 3,
    NULL))
    error ...
```

The NULL at the end of the argument list for all vips functions marks the end of the options.

**Comment by jcupitt** [comment]
There's a chapter in the nip2 manual about image mosaicing, it's very old now, but it might be relevant:

http://www.vips.ecs.soton.ac.uk/supported/current/doc/html/nip2/nipguidese4.html#x11-100003.1

That's for people using a mechanical easel to move a painting under a camera and lights, but a lot of it applies to microscopes too.

**Comment by alonit** [comment]
Good stuff, Thanks.

Yes, overlap is around 20 pixels.

Can you please explain how did you calculate and get 6 for `harea`, and 3 for `hwindow `?

Thanks

**Comment by jcupitt** [comment]
I just turned down harea until it started working. 20 is rather small, are sure there will always be some feature in the overlap it can find?

**Comment by jcupitt** [comment]
You can make better mosaics if you do something like this:

- lay out all images in their approximate positions and extract all overlaps
- search all overlap pairs for suitable tie points and correlate to get a refined overlap position ... match at least three tie points in each and make sure they agree
- you can improve this by doing gradient matching, it can work even without a visible tie-point
- but note that for some overlaps it will not be possible to get a refined overlap, or that the refined overlap will be wrong, and that your set of overlaps are almost certain to be inconsistent
- put all your constraints into a big system of equations and solve with something like LMS to get a set of overlaps which minimise average error
- you can also solve to minimise contrast differences at edges at this stage (`global_balance()` does this)
- now join the array of images pairwise with `vips_merge()`

**Comment by alonit** [comment]
Thanks a lot for this suggestion, I'll check this out.

Meanwhile I wonder what is the ideal file format that the scanned image tiles might be, so that the stitcher will get the most accurate and high quality input to produce the most high quality and accurate stitched image for the final pyramid builder.

I'm taking into account that for pyramid build performance and pyramid based viewer it will be the only one that apply loss compression (.jpeg) so no other station (scanner/stitcher) should apply also loss compression on the same tiles.

I have 32bit process for the scanner that acquire raw color camera images (12bit Bayer format), and I have to pass this images (tiles) to a 64bit process that runs the vips sticher and pyramid builder. (vips is 64bit package).

So I think to save the bayer images format that come from the camera to a standard raw image format such as DNG, but I don't know if vips support this format or any other raw formats to read them from disc for latter stitch/pyramid processing.

I also think about another way: saving the bayer images to tiff (with lossless compression such as LZW), but this will need extra heavy processing, including:
1. extract the RGB channels from the bayer image
2. combine them to a single color image
3. wrap the combined image under bitmap
4. compress with LZW
5. save it to .tif format.  

The sticher that will read those .tif file as tiles.

To sum up the above, my question is: what is the recommended file format for the vips based stitcher that the original color camera images (bayer) should be saved and then read into the vips stitcher, taking accuracy, quality and performance of the final pyramid product?

Thanks

**Comment by jcupitt** [comment]
I think you'll need to demosaic before stitching. You can't run the demosaic after stitching -- the Bayer pattern will be all messed up. I'd think the sequence would need to be:

* acquire, black level correction, shading correction, dead pixel removal
* demosaic to TIFF
* correction for lens distortion and chromatic aberration
* colour correction
* extract approximate overlaps and analyze for tile positions
* blend to output pyramid

I've made (I think?) three scanners like this and used vips for every step. You can make the demosaic yourself, or just use dcraw. I'd focus on getting it working first, and speed second.

**Comment by jcupitt** [comment]
I'll close. Please feel free to open a new issue for any follow-ups.

**Comment by harukizaemon** [comment]
@jcupitt I'm finally turning my attention to stitching again and I note the following suggestions:

>* lay out all images in their approximate positions and extract all overlaps
I have this as a set of NxM rectangles `(x, y, w, h)`.

>* search all overlap pairs for suitable tie points and correlate to get a refined overlap position ... match at least three tie points in each and make sure they agree
Not sure who to go about this. Do I use `im_correl()`?

>* put all your constraints into a big system of equations and solve with something like LMS to get a set of overlaps which minimise average error
Can do this.

>* now join the array of images pairwise with `vips_merge()`
Makes sense

**Comment by harukizaemon** [comment]
@jcupitt I'm finally turning my attention to stitching again I had thought I would do the following:

* Conceptually, layout the tiles NxM
* Use `vips_mosaic` to join tiles from the (approximate) centre out to minimize translation/rotation at the centre
* Profit

Here however I note the following suggestions:

>* lay out all images in their approximate positions and extract all overlaps
I have this as a set of NxM rectanges

>* search all overlap pairs for suitable tie points and correlate to get a refined overlap position ... match at least three tie points in each and make sure they agree
Not sure who to go about this. Do I use

>* put all your constraints into a big system of equations and solve with something like LMS to get a set of overlaps which minimise average error
>* you can also solve to minimise contrast differences at edges at this stage (global_balance() does this)
>* now join the array of images pairwise with `vips_merge()`
Makes sense
